### ISSUE TITLE
(suggestion) Color \iffalse - \fi as comment color

### ISSUE THREAD

**Issue by htna** [texstudio-org/texstudio#618]
In TexStudio, comment can be shown in different color from other regular text colors.

\iffalse-\fi block is also considered as a comment in LaTeX. I wish iffalse - fi is colored as comment as well.

**Comment by muzimuzhi** [comment]
Since TeX is a macro language, directly treat `<case 1>` in block `\iffalse <case 1> \else <case 2> \fi` as comment, without any macro expansion, is risky.

For example, 
```latex
\documentclass{article}
\def\abc{before \fi inside abc \iftrue after}

\begin{document}
\iffalse \abc \fi             % produces nothing, as expected
\expandafter\iffalse \abc \fi % produces "inside abc after"
\end{document}
```

Of course, the above example is rarely used in practice, but only to point out a "potential problem". There are other cases in which the `<case 1>` block should not been treated as comment. And basically, without any macro expansion, it is hardly possible to do semantic analysis on TeX.

**Comment by timhoffm** [comment]
@muzimunzhi You are right about the macro character of TeX. Without the context of the complete preceding document, you can never know for sure what a certain TeX expression will do. A correct macro expansion solution would essentially be a complete TeX processor. That's neither desired nor feasible performance-wise in an editor. We have to use heuristics to interpret TeX files; and that's perfectly fine as long as the heuristics are unlikely to fail (e.g. we assume that `\cite{}` contains a reference. While you could redefine this, it's unlikely to happen in practice.)

The question here is: What are there (realistic and not too rare) use cases for `\iffalse`:
- As the OP mentionend one is block-commenting. - Even though I don't think it's a very compelling case as TXS also supports (un)commenting multiple lines in a selection using `%`; and there is `\begin{comment}\end{comment}` as well.
- Do you have a good case for using `\iffalse` that would be harmed by comment-coloring the block?

**Comment by muzimuzhi** [comment]
> Do you have a good case for using `\iffalse` that would be harmed by comment-coloring the block?

Whether comment-coloring the `\iffalse-\fi` block would be harmed depends on which kinds of files the texstudio will open.

Suppose we divide files into two kinds,
 - **User files**: latex files (mostly are `.tex`) producing actual docs (learning notes, lecture slides, journal papers, or academic theses)
 - **Developer files**: files providing functions, e.g., `.sty`, `.cls`, or even`.dtx`

**In user files**, comment-coloring may rarely be harmed, as normally `\iffalse` would been used to markup block comment.

**In developer files**, things are different since
1. `\iffalse` may occur in `\let`, for example `\let\if@twocolumn\iffalse`.
    - LaTeX2e kernel contains 20+ usages of this kind. 
    - My local in-complete installation of TeX Live 2019 contains 130+ usages of this kind in total. 
    - A more complete example ([link to source](https://github.com/latex3/latex2e/blob/master/base/ltdefns.dtx#L770-L782)):
    ```latex
    \def\newif#1{%
      \count@\escapechar \escapechar\m@ne
        \let#1\iffalse
        \@if#1\iftrue
        \@if#1\iffalse
      \escapechar\count@}
   ```
1. `\iffalse-\fi` block is complete, but the contents between them are not always treated as comment, for exampe in usages of ``\iffalse{\fi\ifnum 0=`}\fi`` (it is used to insert an un-paired brace in tabular/array related macros, see [explanation](https://tex.stackexchange.com/q/188704/79060))
1. `\iffalse`, without the paired `\fi` part, may occur in a macro definition, for example ([link to source](https://github.com/latex3/latex2e/blob/e77bc35853d4c07b3245b08a0b7dd01a5541e778/base/doc.dtx#L2182-L2193))
    ```latex
    \def\environment{\begingroup
       \catcode`\\12
       \MakePrivateLetters \m@cro@ \iffalse}
    ```
1. `\iffalse` may follow and been absorbed by some expandable macros, for example `\string\iffalse`. Macros handling "write to aux files" may use this mechanism.

------

To sum up, 
1. `\iffalse` is rarely used both in user and developer files.
1. In developer files, there are several kinds of usages of `\iffalse` in which simply comment-coloring would be harmed.
1. In user files, instead of `\iffalse`, there are already editor-friendly "block comment" alternatives. (as @timhoffm has mentioned)
1. If considering add "comment-color of `\iffalse`" to texstudio, I suggest implement it as a checkbox configuration item and leave it un-checked by default.

**Comment by timhoffm** [comment]
@muzimuzhi Thanks for digging into this!

Given your summary, I would refrain from implementing coloring of \iffalse. If somebody really wants to color it he/she can simply modify the language definition [as described in the manual](http://texstudio.sourceforge.net/manual/current/usermanual_en.html#LANGUAGEDEF).

**Comment by htna** [comment]
> @muzimunzhi You are right about the macro character of TeX. Without the context of the complete preceding document, you can never know for sure what a certain TeX expression will do. A correct macro expansion solution would essentially be a complete TeX processor. That's neither desired nor feasible performance-wise in an editor. We have to use heuristics to interpret TeX files; and that's perfectly fine as long as the heuristics are unlikely to fail (e.g. we assume that `\cite{}` contains a reference. While you could redefine this, it's unlikely to happen in practice.)
> 
> The question here is: What are there (realistic and not too rare) use cases for `\iffalse`:
> 
> * As the OP mentionend one is block-commenting. - Even though I don't think it's a very compelling case as TXS also supports (un)commenting multiple lines in a selection using `%`; and there is `\begin{comment}\end{comment}` as well.
> * Do you have a good case for using `\iffalse` that would be harmed by comment-coloring the block?

How can I specify color of block by \begin{comment} ... \end{comment}?
It seems that the block is colored as verbatim.
